### ISSUE TITLE
Should the batch size for each rank is the same? Any support bucketing with different batch sizes?

### ISSUE THREAD

**Issue by XuesongYang** [horovod/horovod#1060]
To my knowledge so far, the batch size for each rank should keep the same, such that current framework does not support bucketing by length with different batch sizes. Is my understanding correct? Thanks

**Comment by tgaddair** [comment]
Hey @XuesongYang, yes the batch sizes should be the same across ranks.  The way the allreduce operation works, we need tensors to be the same size so everything aligns correctly.  In theory, you could pad the gradients, but that may not give you the desired effect (the gradients would be skewed towards 0).  The way we typically scale the learning rate would also be affected by varying the batch sizes across the workers.

Could you describe in a bit more detail what you're trying to achieve?  Is there a paper that discusses the approach you're exploring?

**Comment by Young768** [comment]
Hi @tgaddair , I also wonder if Horovod supports adaptive/dynamic batch size during training. These are related paper I have found: [[1]](http://proceedings.mlr.press/v54/de17a.html), [[2]](https://arxiv.org/abs/1712.02029)

**Comment by tgaddair** [comment]
Hey @Young768, as long as the batch size changes with the same policy on every worker, it should be fine (for example: epoch 1 = 32, epoch 2 = 64, etc.).

**Comment by Young768** [comment]
Hi @tgaddair How about the same batch, but different workers have different batch size? Their tensor size is the same, which ensure allreduce operations.

**Comment by Young768** [comment]
Hi @tgaddair could you please provide these related APIs you have mentioned that can change batch size between different epochs on Horovod?

**Comment by jdaaph** [comment]
> Hey @XuesongYang, yes the batch sizes should be the same across ranks. The way the allreduce operation works, we need tensors to be the same size so everything aligns correctly. In theory, you could pad the gradients, but that may not give you the desired effect (the gradients would be skewed towards 0). The way we typically scale the learning rate would also be affected by varying the batch sizes across the workers.
> 
> Could you describe in a bit more detail what you're trying to achieve? Is there a paper that discusses the approach you're exploring?

Thanks for the clarification @tgaddair , however in my own testing of a NLP training task, I found that even with different batch sizes for each rank I can still run my program without any error or warning. Would you say that Horovod now would give undefined behavior and calculate something wrong even if there is no error reported? Thank you!

**Comment by XuesongYang** [comment]
> Hey @XuesongYang, yes the batch sizes should be the same across ranks. The way the allreduce operation works, we need tensors to be the same size so everything aligns correctly. In theory, you could pad the gradients, but that may not give you the desired effect (the gradients would be skewed towards 0). The way we typically scale the learning rate would also be affected by varying the batch sizes across the workers.
> 
> Could you describe in a bit more detail what you're trying to achieve? Is there a paper that discusses the approach you're exploring?

I just realized that I forgot to reply this message. Typically for the use case of speech recognition, we usually organize a batch with a constant number that equals to the multiplication of number of utterances in a batch and the max length (in frames) of the utterances in a batch.

In this case, does Horovod correctly support this setup? I can run experiments through without any errors so far, but I am just wondering if this setup is correct. Thanks.

**Comment by formath** [comment]
I start horovod with two workers in which the first one has 100 training data and the second one has 90 training data. While setting `batch_size=10`, the first worker will run `session.run(train_op)` 10 times while the second one runs 9 times. So in the latst batch of the first worker, `allreduce` will not accept any gradient from the second worker. Does horovod handle this problem?
